Simplify lql init to rely on template questions, clean up version checks, and deduplicate CLI test helpers

sdqede opened the pull request:
# ✨ Summary

This PR:

- ✅ Makes `lql init` rely entirely on template `.questions.json` + **inquirerer** for identity questions (name, email, license, repo name, etc.).
- 🧹 Cleans up ALL‑CAPS keys and duplicated init logic in CLI tests and init flows.
- 🔀 Separates **pgpm** vs **@launchql/cli** update‑check state while sharing a common mechanism.
- 🧱 Improves template caching robustness and `create-gen-app` license template handling.

---

## 1️⃣ Init flows are now template‑driven

### `packages/pgpm/src/commands/init/workspace.ts`

Workspace init now:

- 📝 Prompts only for:
  - `name` (workspace name).
- 🚪 Closes the **Inquirerer** instance before `create-gen-app` runs to avoid double‑echo keystrokes.
- 📤 Delegates everything else (full name, email, license, `repoName`, `username`, etc.) to:
  - `pgpm-boilerplates/workspace/.questions.json`
  - inquirerer `defaultFrom` resolvers (e.g. `git.user.name`, `git.user.email`, `npm.whoami`).
- 🧹 Removes:
  - `getGitConfigInfo` usage.
  - `extraQuestions` and any special‑cased `repoName` / license questions.
  - Any **"LaunchQL User"** default.

### `packages/pgpm/src/commands/init/module.ts`

Module init now:

- 📝 Prompts only for:
  - `moduleName` (text).
  - `extensions` (checkbox, options from `getAvailableModules()`, with `allowCustomOptions: true`).
- 🚪 Closes **Inquirerer** before calling `initModule`.
- 📦 Passes a clean, camelCase answers object into `LaunchQLPackage.initModule`:
  - `moduleName`, `description`, `author`, `extensions`, `packageIdentifier`.
- 🧹 Removes all ALL‑CAPS answer keys and custom `username` / `email` logic; those now come solely from the module template’s `.questions.json`.

---

## 2️⃣ Template scaffolding & cache validation

### `packages/core/src/core/template-scaffold.ts`

Template cache handling is now stricter and safer:

- ✅ Validates a cached template repo before use:
  - **Workspace templates** require:
    - `workspace/pgpm.json`
    - `workspace/package.json`
  - **Module templates** require:
    - `module/pgpm.plan`
    - `module/package.json`
- ♻️ If the cache is incomplete or invalid, we:
  - Clear that cache directory.
  - Reclones from `https://github.com/launchql/pgpm-boilerplates.git` using:

```bash
git clone --depth 1 --single-branch https://github.com/launchql/pgpm-boilerplates.git
```

- 🛡️ This prevents Jest (or a killed run) from leaving a broken cache that would previously cause:
  - `Not inside a workspace` errors.
  - Empty / inconsistent snapshots.

---

## 3️⃣ Update checks: shared logic, separate state

Shared logic for **pgpm** and **CLI** lives in:

- `packages/pgpm/src/utils/update-check.ts`
- `packages/pgpm/src/utils/update-config.ts`
- `packages/types/src/update.ts`

### 🔧 Config structure

```ts
interface UpdateCheckConfig {
  lastCheckedAt: number;
  latestKnownVersion: string;
  lastPromptedVersion?: string;
}
```

- ⏱ TTL: `UPDATE_CHECK_TTL_MS = 7 days`.

### 🔁 Behavior

- Read existing config and decide whether to hit npm (`shouldCheck`).
- On TTL expiry or missing config:
  - Fetch latest version from npm.
  - Write:

    ```json
    {
      "lastCheckedAt": <now>,
      "latestKnownVersion": "<npm-version>",
      "lastPromptedVersion": "<npm-version>" | undefined
    }
    ```

- When current version is outdated and `lastPromptedVersion !== latestKnownVersion`:
  - ⚠️ Log a one‑time upgrade warning.
  - Update config with `lastPromptedVersion = latestKnownVersion`.
- Otherwise, the config is left untouched.

### 📁 pgpm config path

From `packages/types/src/update.ts`:

```ts
export const UPDATE_CHECK_APPSTASH_KEY = 'pgpm-update-check';
export const UPDATE_PACKAGE_NAME = 'pgpm';
```

From `packages/pgpm/src/utils/update-config.ts`:

```ts
const toolName = options.toolName ?? 'pgpm';
const fileName = `${(options.key ?? UPDATE_CHECK_APPSTASH_KEY)
  .replace(/[^a-z0-9-_]/gi, '_')}.json`;
// → ~/.pgpm/config/pgpm-update-check.json
```

📌 Result:

- **pgpm** update config lives at:

```text
~/.pgpm/config/pgpm-update-check.json
```

### 📁 CLI (@launchql/cli) config path

From `packages/cli/src/commands.ts`:

```ts
const pkg = readAndParsePackageJson();
await checkForUpdates({
  command,
  pkgName: pkg.name,     // '@launchql/cli'
  pkgVersion: pkg.version,
  toolName: 'lql',       // directory namespace
  key: pkg.name          // '@launchql/cli'
});
```

`update-config.getConfigPath` normalizes the key:

```ts
const toolName = options.toolName ?? 'pgpm';
const fileName = `${(options.key ?? UPDATE_CHECK_APPSTASH_KEY)
  .replace(/[^a-z0-9-_]/gi, '_')}.json`;
```

- `toolName: 'lql'` → appstash dir → `~/.lql/config/`
- `key: '@launchql/cli'` → normalized to `_launchql_cli`

📌 Final path:

```text
~/.lql/config/_launchql_cli.json
```

### 🤔 Why the naming difference?

- **pgpm** is older and originally had a one‑off update check, so it used a fixed name: `'pgpm-update-check'`.
- We keep this to avoid breaking existing user configs in `~/.pgpm/config/pgpm-update-check.json`.
- For **@launchql/cli**, we intentionally key the config by npm package name:
  - Future‑proofs the helper so multiple CLIs can coexist without collisions.
  - Each package gets its own config file under its own `toolName` namespace (here, `lql`).

Under the hood, both use the same pattern:

```text
~/.<toolName>/config/<normalized key>.json
```

- **pgpm**: `toolName = 'pgpm'`, `key = 'pgpm-update-check'`
  - → `~/.pgpm/config/pgpm-update-check.json`
- **CLI**: `toolName = 'lql'`, `key = '@launchql/cli'`
  - → `~/.lql/config/_launchql_cli.json`

We can unify this later if we introduce a migration path, but for now this keeps existing pgpm configs working while giving the CLI a clean, package‑scoped cache.

---

## 4️⃣ CLI tests: deduplicated init defaults

### `packages/cli/test-utils/init-argv.ts`

New helpers:

```ts
export const addInitDefaults = (argv: ParsedArgs): ParsedArgs => {
  const baseName =
    (argv.moduleName as string) ||
    (argv.name as string) ||
    'module';

  const defaults = {
    fullName: 'Tester',
    email: 'tester@example.com',
    moduleName: argv.workspace ? 'starter-module' : baseName,
    username: 'tester',
    repoName: baseName,
    license: 'MIT',
    access: 'public',
    packageIdentifier: baseName,
    moduleDesc: baseName
  };

  return { ...defaults, ...argv };
};

export const withInitDefaults = (
  argv: ParsedArgs,
  defaultRepo = 'https://github.com/launchql/pgpm-boilerplates.git'
): ParsedArgs => {
  const args = addInitDefaults(argv);
  if (!Array.isArray(args._) || !args._.includes('init')) return args;

  return {
    ...args,
    repo: args.repo ?? defaultRepo,
    templatePath: args.templatePath ??
      (args.workspace ? 'workspace' : 'module')
  };
};
```

- Exported via `packages/cli/test-utils/index.ts`.

### `packages/cli/test-utils/fixtures.ts`

`TestFixture.runCmd` now:

```ts
if (Array.isArray(argv._) && argv._.includes('init')) {
  argv = withInitDefaults(argv, DEFAULT_REPO);
}
```

- Removes the inline defaults block and keeps all init defaults in one place.

### Tests using the helper

- `packages/cli/__tests__/init.test.ts`
  - Imports `withInitDefaults` from `../test-utils`.
  - Removes local `addDefaults` / `withInitDefaults` and the `DEFAULT_REPO` constant.
  - All init scenarios go through the shared helper.

- `packages/cli/__tests__/extensions.test.ts`
  - Uses `withInitDefaults(argv)` when running `init` as part of the extension flow.
  - No more ALL‑CAPS argv keys or repeated `||` chains in this test.

---

## 5️⃣ Boilerplates and license templates

- 🗂 The CLI no longer uses a local `./launchql/boilerplates` directory.
  - Both runtime and tests rely on the remote **pgpm-boilerplates** repo, aligning with `init-command.plan.md`.

### 📜 Workspace template license question

`pgpm-boilerplates/workspace/.questions.json` now includes the same license selection as the module template:

```json
{
  "name": "____license____",
  "message": "Choose a license",
  "type": "list",
  "options": ["MIT", "Apache-2.0", "ISC"],
  "required": true
}
```

### 📦 `dev-utils/packages/create-gen-app/package.json`

- Adjusted the copy script so license templates are copied into:

```text
dist/licenses-templates/
```

- This allows `create-gen-app`’s license resolver to find `APACHE-2.0.txt`, etc., and eliminates the:

> `License "Apache-2.0" is not supported`

warning when templates are present.

---

## ✅ Manual testing

All commands run locally from `./launchql` using `pnpm`.

### 🧪 CLI tests from a clean pgpm cache

```bash
rm -rf ~/.pgpm
pnpm --filter @launchql/cli test
```

Confirms:

- `pgpm` clones `https://github.com/launchql/pgpm-boilerplates.git` (with `--depth 1 --single-branch`) into the cache.
- All `@launchql/cli` Jest suites pass **without** requiring `--runInBand`.
- Cached repo is valid and reused on subsequent runs.

### 📖 CLI help

```bash
node packages/cli/dist/index.js init -h
```

Displays:

- `--workspace` flag for workspace init.
- `--repo`, `--template-path`, `--from-branch` options.

Matches README behavior:

- `lql init --workspace` → create workspace.
- `lql init` inside workspace → create module.

### 🔢 Version behavior and config

```bash
npm view @launchql/cli version
# confirm against packages/cli/dist/package.json

vim ~/.lql/config/_launchql_cli.json
cat ~/.lql/config/_launchql_cli.json
```

Example config:

```json
{
  "lastCheckedAt": 1764828956271,
  "latestKnownVersion": "4.17.17",
  "lastPromptedVersion": "4.17.17"
}
```

Verified:

- When `now - lastCheckedAt > 7 days`:
  - Next CLI invocation re‑checks npm, refreshes `latestKnownVersion`, and sets `lastCheckedAt = now`.
- When `latestKnownVersion > current CLI version` **and** `lastPromptedVersion` is different:
  - CLI prints a one‑time upgrade warning and updates `lastPromptedVersion` to the latest version (and `lastCheckedAt = now`).
- Otherwise, config stays untouched.

### 🧪 Workspace and module scaffolding via CLI

#### Workspace

```bash
rm -rf /tmp/lql-test
node packages/cli/dist/index.js init   --cwd /tmp/lql-test   --workspace
```

Creates `/tmp/lql-test/<workspace-name>/` with:

- `package.json`
- `pgpm.json`

All identity prompts (name, email, license, `repoName`, `username`, etc.) come from the workspace template’s `.questions.json` and inquirerer.

#### Module (inside workspace)

```bash
node packages/cli/dist/index.js init   --cwd /tmp/lql-test/<workspace-name>
```

Creates `/tmp/lql-test/<workspace-name>/packages/<module-name>/` with:

- `package.json`
- `pgpm.plan`

Extension choices and module metadata are entirely template‑driven; no extra prompts or git lookups in the CLI itself.

sdqede committed:
refactor: remove @launchql/templatizer and integrate create-gen-app for template scaffolding

- Deleted the @launchql/templatizer package and its related files.
- Updated CLI commands to use create-gen-app for scaffolding modules and workspaces.
- Introduced caching for templates with a default repository.
- Enhanced documentation to reflect changes in template usage and commands.
- Cleaned up boilerplate files and configurations across the workspace.

sdqede committed:
feat: enhance update check in CLI commands

- Added toolName and key parameters to the update check function in CLI commands.
- Improved the update check functionality for better tracking of package updates.

sdqede committed:
feat: prompt for user identity if Git config is missing

- Updated the runModuleSetup and runWorkspaceSetup functions to prompt users for their full name and email address if not found in the Git configuration.
- Refactored getGitConfigInfo to ensure it returns empty strings for username and email if retrieval fails, improving error handling.

sdqede committed:
feat: enhance module and workspace setup with additional prompts

- Updated runModuleSetup to include packageIdentifier as an optional argument.
- Enhanced runWorkspaceSetup to prompt for repoName and license if not provided, ensuring all necessary information is collected upfront.
- Disabled interactive prompts in both functions to streamline the setup process and avoid nested TTY handling.

pyramation commented on packages/cli/__tests__/extensions.test.ts:
```diff
@@ -29,6 +33,24 @@ describe('cmds:extension', () => {
+        license: 'MIT',
+        access: 'public',
+        packageIdentifier: (argv.name as string) || (argv.MODULENAME as string) || 'module',
+        moduleDesc: (argv.name as string) || (argv.MODULENAME as string) || 'module'
```
make it simple, just use very simple defaults and make code more readable

pyramation commented on packages/cli/__tests__/init.test.ts:
```diff
@@ -9,6 +12,29 @@ import { commands } from '../src/commands';
+  const defaults = {
+    fullName: 'Tester',
+    email: 'tester@example.com',
+    moduleName: argv.workspace ? 'starter-module' : argv.name || argv.MODULENAME || 'module',
```
again, same comment as before. THERE IS NO REASON FOR CAPS.

There is no reason to have all the `||`

pyramation commented on packages/cli/test-utils/fixtures.ts:
```diff
@@ -53,6 +54,25 @@ export class TestFixture {
+      const defaults = {
+        fullName: 'Tester',
+        email: 'tester@example.com',
+        moduleName: argv.workspace ? 'starter-module' : argv.name || argv.MODULENAME || 'module',
```
this code looks copy and pasted.... let's deduplicate all 3 (or more) of the places this code is used, please.

pyramation commented on packages/pgpm/src/commands/init/module.ts:
```diff
@@ -48,40 +46,74 @@ export default async function runModuleSetup(
   const answers = await prompter.prompt(argv, moduleQuestions);
   const modName = sluggify(answers.MODULENAME);
```
why `MODULENAME` still

pyramation commented on packages/pgpm/src/commands/init/module.ts:
```diff
@@ -48,40 +46,74 @@ export default async function runModuleSetup(
-    name: modName,
-    // @ts-ignore
     USERFULLNAME: username,
     USEREMAIL: email,
```
more `CAPS`

USERFULLNAME
USEREMAIL

pyramation commented on packages/pgpm/src/commands/init/workspace.ts:
```diff
@@ -17,42 +15,104 @@ export default async function runWorkspaceSetup(
+
+  let { email, username } = getGitConfigInfo();
+
+  if (!username || !email) {
```
why are we doing this? we should not need to ask the user for this?

sdqede replied:
I noticed that when the user hasn’t set a global git config, lql throws an error. To address this, I added a check: if the global git config is missing, we prompt the user to enter the values manually; if it’s already set, we keep the original behaviour and use the existing git config.
Should we continue throwing an error when the global git config is missing, or handle it gracefully with the prompt?

pyramation commented on packages/pgpm/src/commands/init/workspace.ts:
```diff
@@ -17,42 +15,104 @@ export default async function runWorkspaceSetup(
+      ...answers,
+      ...extra,
+      workspaceName: answers.name,
+      fullName: username || email || 'LaunchQL User',
```
replace any and all "LaunchQL User" — just put "user" if that. DO this everywhere you have this.

pyramation commented:
@sdqede we're getting close, much better. Still some work to do:

- remember, do NOT call getGitConfigInfo or try to get git info, let inqurierer handle that now with it's resolvers. https://github.com/launchql/pgpm-boilerplates/blob/968f5d6c7be0880252a6dce6d97837a486d683f9/workspace/.questions.json#L12

- there is way too much ALL CAPS stuff, remove it all. all of it.

- way to many extra questions or the creation of questions... I thought the entire point was to just parse the templates, and get the questions from .questions and the templates, and then ask user via inqurierer?  In other words, DO NOT create questions or try to guess any propery names like license, email, user, username, or ask for those. JUST USE THE TOOLS WE WROTE from create-gen-app

thank you @sdqede :)

sdqede committed:
Normalize init args to camelCase and update snapshots

sdqede committed:
refactor: streamline module and workspace setup by removing unnecessary prompts

- Removed prompts for user identity (username and email) in runModuleSetup and runWorkspaceSetup functions, simplifying the setup process.
- Updated snapshot tests to reflect changes in author information and repository URLs.
- Closed prompter before template prompts to prevent overlapping input handling.

sdqede committed:
fix(core): validate and repair cached template repos

Add isValidTemplateRepo in packages/core/src/core/template-scaffold.ts to verify that cached boilerplate clones actually contain a usable template (workspace/pgpm.json + package.json or module/pgpm.plan + package.json).
Change scaffoldTemplate to treat a cache entry as valid only when metadata is fresh and isValidTemplateRepo returns true; otherwise clear the cache key and reclone from the remote repo with depth: 1 and singleBranch: true.
Fixes intermittent pgpm init / lql init failures where Jest or previous runs left a partially-populated cache directory (e.g. only .github/), causing workspace detection to fail and tests to error with “You must be in a LaunchQL workspace” and empty file snapshots.
Ensures pnpm test --filter @launchql/cli passes reliably even after deleting ~/.pgpm or the Jest-specific cache, and makes CLI users automatically recover from corrupted template cache state without manual intervention.

sdqede committed:
refactor(cli): simplify default argument assignment in extension setup

- Introduced a baseName variable to streamline the assignment of moduleName, repoName, packageIdentifier, and moduleDesc in the extension setup process.
- This change enhances code readability and maintainability by reducing redundancy in default value assignments.

sdqede committed:
refactor(cli): streamline default argument assignment in init tests

- Introduced a baseName variable to simplify the assignment of moduleName, repoName, packageIdentifier, and moduleDesc in the init test setup.
- This change improves code readability and reduces redundancy in default value assignments.

sdqede committed:
refactor(cli): consolidate default argument handling in tests

- Replaced manual default argument assignments with a centralized `withInitDefaults` function in both extension and init tests.
- This change enhances code clarity and maintainability by reducing redundancy in default value assignments across test files.

sdqede committed:
chore: update create-gen-app version to 0.3.1 in package.json and pnpm-lock.yaml

- Bumped create-gen-app dependency version from 0.3.0 to 0.3.1 in core and pgpm package.json files.
- Updated pnpm-lock.yaml to reflect the new version and its integrity hash.

sdqede committed:
chore(tests): update snapshots to include LICENSE file in extension and init tests

- Added LICENSE file to the snapshots for both extension and init tests to ensure accurate representation of project structure.
